File: impl/csharp/test/IdentityRecordsKatTest.cs
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.Json;

using Naalp;
using Xunit;

namespace Sh.Bubblefish.Naalp.Tests
{
    /// <summary>
    /// The ENVELOPE-adjacent identity RECORD + THREAD surfaces (design.md §5.3/§5.4/§5.2, R-1.4):
    /// <see cref="Identity.RevocationRecord"/>, <see cref="Identity.RevokedAt"/>,
    /// <see cref="Identity.VerifyRevocation"/>, <see cref="Identity.ForeignLinkRecord"/>,
    /// <see cref="Identity.VerifyForeignLink"/>, <see cref="Identity.RotationEvidence"/>,
    /// <see cref="Identity.Thread"/>, <see cref="Identity.Thread.Attributable"/>,
    /// <see cref="Identity.ResolveThread"/> -- graded against the independent, non-circular oracle
    /// (tools/identity_records_oracle.py -&gt; vectors/identity_records/cases.json), i.e.
    /// C# == Go == Rust == oracle. Mirrors impl/go/identity/identity_records_oracle_test.go and
    /// impl/rust/src/identity.rs's identity_records_oracle.py section.
    ///
    /// <para>SECURITY-CRITICAL fail-closed anchor: <see cref="Identity.VerifyRevocation"/> checks
    /// signer-id MEMBERSHIP (record.key OR a deployer-configured recovery id) BEFORE the signature.
    /// "recovery_key_not_configured_reject" is the mutation anchor -- a valid recovery-key signature
    /// presented against an EMPTY authorized-recovery-id set MUST be rejected SignerMismatch; dropping
    /// the membership guard (authorized := true unconditionally) flips it to accept.</para>
    ///
    /// <para>Run: <c>dotnet test test/Bubblefish.Naalp.Tests.csproj -c Release --filter FullyQualifiedName~IdentityRecordsKatTest</c></para>
    /// </summary>
    public sealed class IdentityRecordsKatTest
    {
        private static string? FindVector()
        {
            var d = new DirectoryInfo(AppContext.BaseDirectory);
            for (int i = 0; i < 10 && d != null; i++)
            {
                string p = Path.Combine(d.FullName, "vectors", "identity_records", "cases.json");
                if (File.Exists(p))
                {
                    return p;
                }
                d = d.Parent;
            }
            return null;
        }

        private static JsonElement LoadCorpus(out JsonDocument doc)
        {
            string? p = FindVector();
            Assert.True(p != null, "committed identity_records vector not found from " + AppContext.BaseDirectory);
            doc = JsonDocument.Parse(File.ReadAllText(p!, Encoding.UTF8));
            return doc.RootElement;
        }

        private static byte[] Hex(string s) => Convert.FromHexString(s);
        private static string HexLower(byte[] b) => Convert.ToHexString(b).ToLowerInvariant();

        // GetU64 reads a JSON integer that may exceed long.MaxValue and returns it as the SAME BIT
        // PATTERN in a signed long -- the wire convention this port already uses for Cbor.U.V.
        private static long GetU64(JsonElement el, string prop)
        {
            JsonElement v = el.GetProperty(prop);
            if (v.TryGetInt64(out long l))
            {
                return l;
            }
            return unchecked((long)v.GetUInt64());
        }

        private static long? GetNullableU64(JsonElement el, string prop)
        {
            if (!el.TryGetProperty(prop, out JsonElement v) || v.ValueKind == JsonValueKind.Null)
            {
                return null;
            }
            if (v.TryGetInt64(out long l))
            {
                return l;
            }
            return unchecked((long)v.GetUInt64());
        }

        private static List<string> StringList(JsonElement arr)
        {
            var outp = new List<string>();
            foreach (JsonElement e in arr.EnumerateArray())
            {
                outp.Add(e.GetString()!);
            }
            return outp;
        }

        // ---- RevocationRecord.Bytes (§5.3) -----------------------------------------------------

        [Fact]
        public void RevocationRecordBytesMatchesOracle()
        {
            JsonElement root = LoadCorpus(out JsonDocument doc);
            using (doc)
            {
                JsonElement cases = root.GetProperty("revocation").GetProperty("record_bytes");
                int n = 0;
                foreach (JsonElement tc in cases.EnumerateArray())
                {
                    n++;
                    string name = tc.GetProperty("name").GetString()!;
                    var r = new Identity.RevocationRecord(tc.GetProperty("key").GetString()!, GetU64(tc, "not_after"));
                    string want = tc.GetProperty("bytes_hex").GetString()!;
                    Assert.True(want == HexLower(r.Bytes()), name);
                }
                Assert.True(n > 0, "no revocation.record_bytes cases loaded");
            }
        }

        // ---- RevokedAt (§5.3) -- MUTATION ANCHOR: "at_boundary_still_valid" pins `>` vs `>=`. -------

        [Fact]
        public void RevokedAtMatchesOracle()
        {
            JsonElement root = LoadCorpus(out JsonDocument doc);
            using (doc)
            {
                JsonElement scenarios = root.GetProperty("revocation").GetProperty("revoked_at");
                int n = 0;
                foreach (JsonElement sc in scenarios.EnumerateArray())
                {
                    n++;
                    string name = sc.GetProperty("name").GetString()!;
                    string queryKey = sc.GetProperty("query_key").GetString()!;
                    long queryPosition = GetU64(sc, "query_position");
                    bool expectRevoked = sc.GetProperty("expect_revoked").GetBoolean();
                    long? expectNotAfter = GetNullableU64(sc, "expect_not_after");

                    bool revoked = false;
                    long notAfter = 0;
                    foreach (JsonElement rv in sc.GetProperty("revocations").EnumerateArray())
                    {
                        string key = rv.GetProperty("key").GetString()!;
                        if (key != queryKey)
                        {
                            continue;
                        }
                        long na = GetU64(rv, "not_after");
                        var rec = new Identity.RevocationRecord(key, na);
                        if (Identity.RevokedAt(rec, queryPosition))
                        {
                            revoked = true;
                            notAfter = na;
                        }
                    }
                    Assert.True(revoked == expectRevoked, name);
                    if (expectRevoked)
                    {
                        Assert.True(expectNotAfter.HasValue && notAfter == expectNotAfter.Value, name);
                    }
                }
                Assert.True(n > 0, "no revocation.revoked_at scenarios loaded");
            }
        }

        // ---- VerifyRevocation (§5.3, §5.5) -- SECURITY-CRITICAL fail-closed authorization ---------
        // §5.3 permits a Revocation to be signed by the key it revokes OR by a deployer-configured
        // recovery key. VerifyRevocation takes the deployer's authorized recovery-id set and accepts a
        // signer iff its recomputed id equals record.Key or is a member of that set, THEN verifies the
        // signature (membership BEFORE signature, fail-closed). MUTATION ANCHORS:
        // "recovery_key_not_configured_reject" (a valid recovery-key signature with an EMPTY
        // authorized set -> SignerMismatch) and "wrong_key_reject" -- dropping the membership guard
        // flips both to accept.

        [Fact]
        public void VerifyRevocationMatchesOracle()
        {
            JsonElement root = LoadCorpus(out JsonDocument doc);
            using (doc)
            {
                JsonElement cases = root.GetProperty("revocation").GetProperty("verify");
                int n = 0;
                foreach (JsonElement tc in cases.EnumerateArray())
                {
                    n++;
                    string name = tc.GetProperty("name").GetString()!;
                    JsonElement recEl = tc.GetProperty("record");
                    var rec = new Identity.RevocationRecord(recEl.GetProperty("key").GetString()!, GetU64(recEl, "not_after"));
                    int alg = (int)tc.GetProperty("candidate_alg").GetInt64();
                    byte[] pub = Hex(tc.GetProperty("candidate_pubkey_hex").GetString()!);
                    byte[] sig = Hex(tc.GetProperty("sig_hex").GetString()!);
                    List<string> recoveryIds = StringList(tc.GetProperty("authorized_recovery_ids"));
                    bool expectValid = tc.GetProperty("expect_valid").GetBoolean();
                    string expectKind = tc.TryGetProperty("expect_error_kind", out JsonElement ek) ? ek.GetString() ?? "" : "";

                    if (expectValid)
                    {
                        Identity.VerifyRevocation(rec, alg, pub, sig, recoveryIds); // must not throw
                    }
                    else
                    {
                        NaalpException ex = Assert.Throws<NaalpException>(() =>
                            Identity.VerifyRevocation(rec, alg, pub, sig, recoveryIds));
                        if (expectKind.Length > 0)
                        {
                            Assert.True(expectKind == ex.Kind, name + ": got " + ex.Kind);
                        }
                    }
                }
                Assert.True(n >= 7, "expected at least 7 revocation.verify cases, got " + n);
            }
        }

        /// <summary>
        /// The fail-closed anchor, isolated: a VALID recovery-key signature presented against an
        /// EMPTY authorized-recovery-id set MUST be rejected SignerMismatch (§5.5) -- the membership
        /// check runs before the signature check and an unconfigured recovery key confers no
        /// authority. This is the assertion the required mutation witness targets.
        /// </summary>
        [Fact]
        public void RecoveryKeyNotConfiguredIsRejectedFailClosed()
        {
            JsonElement root = LoadCorpus(out JsonDocument doc);
            using (doc)
            {
                JsonElement cases = root.GetProperty("revocation").GetProperty("verify");
                bool found = false;
                foreach (JsonElement tc in cases.EnumerateArray())
                {
                    if (tc.GetProperty("name").GetString() != "recovery_key_not_configured_reject")
                    {
                        continue;
                    }
                    found = true;
                    JsonElement recEl = tc.GetProperty("record");
                    var rec = new Identity.RevocationRecord(recEl.GetProperty("key").GetString()!, GetU64(recEl, "not_after"));
                    int alg = (int)tc.GetProperty("candidate_alg").GetInt64();
                    byte[] pub = Hex(tc.GetProperty("candidate_pubkey_hex").GetString()!);
                    byte[] sig = Hex(tc.GetProperty("sig_hex").GetString()!);
                    List<string> recoveryIds = StringList(tc.GetProperty("authorized_recovery_ids"));
                    Assert.Empty(recoveryIds); // the anchor requires an EMPTY authorized set
                    NaalpException ex = Assert.Throws<NaalpException>(() =>
                        Identity.VerifyRevocation(rec, alg, pub, sig, recoveryIds));
                    Assert.Equal("SignerMismatch", ex.Kind);
                }
                Assert.True(found, "recovery_key_not_configured_reject case not found in the corpus");
            }
        }

        // ---- ForeignLinkRecord.Bytes (§5.4) -- MUTATION ANCHOR: collapsing NFC/NFD to the same bytes
        //      would flip the not-equal assertion below. ------------------------------------------

        [Fact]
        public void ForeignLinkRecordBytesMatchesOracle()
        {
            JsonElement root = LoadCorpus(out JsonDocument doc);
            using (doc)
            {
                JsonElement cases = root.GetProperty("foreign_link").GetProperty("record_bytes");
                var seen = new Dictionary<string, string>();
                int n = 0;
                foreach (JsonElement tc in cases.EnumerateArray())
                {
                    n++;
                    string name = tc.GetProperty("name").GetString()!;
                    var r = new Identity.ForeignLinkRecord(
                        tc.GetProperty("controls").GetString()!,
                        tc.GetProperty("foreign_id").GetString()!,
                        GetU64(tc, "not_after"));
                    string want = tc.GetProperty("bytes_hex").GetString()!;
                    string got = HexLower(r.Bytes());
                    Assert.True(want == got, name);
                    seen[name] = got;
                }
                Assert.True(n >= 2, "expected at least 2 foreign_link.record_bytes cases, got " + n);
                Assert.NotEqual(seen["nfc_form"], seen["nfd_form_different_bytes"]);
            }
        }

        // ---- VerifyForeignLink (§5.4, §5.5) -----------------------------------------------------
        // Valid+unexpired, the not_after boundary (MUTATION ANCHOR for `now > NotAfter`), expiry
        // (ignored, no error), wrong-key (ignored, no error -- the SAME bucket as expiry per §5.5),
        // and non-NFC foreign_id (NonNFC, checked before expiry/signature).

        [Fact]
        public void VerifyForeignLinkMatchesOracle()
        {
            JsonElement root = LoadCorpus(out JsonDocument doc);
            using (doc)
            {
                JsonElement cases = root.GetProperty("foreign_link").GetProperty("verify");
                int n = 0;
                foreach (JsonElement tc in cases.EnumerateArray())
                {
                    n++;
                    string name = tc.GetProperty("name").GetString()!;
                    JsonElement recEl = tc.GetProperty("record");
                    var rec = new Identity.ForeignLinkRecord(
                        recEl.GetProperty("controls").GetString()!,
                        recEl.GetProperty("foreign_id").GetString()!,
                        GetU64(recEl, "not_after"));
                    int alg = (int)tc.GetProperty("candidate_alg").GetInt64();
                    byte[] pub = Hex(tc.GetProperty("candidate_pubkey_hex").GetString()!);
                    byte[] sig = Hex(tc.GetProperty("sig_hex").GetString()!);
                    long now = GetU64(tc, "now");
                    string expectKind = tc.TryGetProperty("expect_error_kind", out JsonElement ek) ? ek.GetString() ?? "" : "";

                    if (expectKind.Length > 0)
                    {
                        NaalpException ex = Assert.Throws<NaalpException>(() =>
                            Identity.VerifyForeignLink(rec, alg, pub, sig, now));
                        Assert.True(expectKind == ex.Kind, name + ": got " + ex.Kind);
                        continue;
                    }

                    bool linked = Identity.VerifyForeignLink(rec, alg, pub, sig, now);
                    bool expectLinked = tc.GetProperty("expect_linked").GetBoolean();
                    Assert.True(linked == expectLinked, name);
                    if (expectLinked)
                    {
                        Assert.True(rec.Controls == tc.GetProperty("expect_controls").GetString(), name);
                        Assert.True(rec.ForeignId == tc.GetProperty("expect_foreign_id").GetString(), name);
                    }
                }
                Assert.True(n > 0, "no foreign_link.verify cases loaded");
            }
        }

        // ---- RotationEvidence / Thread / ResolveThread (§5.2, R-1.4) ---------------------------

        private static List<Identity.RotationEvidence> BuildEvidence(JsonElement evsJson)
        {
            var outp = new List<Identity.RotationEvidence>();
            foreach (JsonElement e in evsJson.EnumerateArray())
            {
                var rec = new Identity.RotationRecord(
                    e.GetProperty("old").GetString()!,
                    e.GetProperty("new").GetString()!,
                    GetU64(e, "not_before"));
                string wantBytes = e.GetProperty("record_bytes_hex").GetString()!;
                Assert.True(wantBytes == HexLower(rec.Bytes()), "RotationRecord.Bytes() (RotationEvidence input)");
                int oldAlg = (int)e.GetProperty("old_alg").GetInt64();
                int newAlg = (int)e.GetProperty("new_alg").GetInt64();
                byte[] oldPub = Hex(e.GetProperty("old_pubkey_hex").GetString()!);
                byte[] newPub = Hex(e.GetProperty("new_pubkey_hex").GetString()!);
                byte[] oldSig = Hex(e.GetProperty("old_sig_hex").GetString()!);
                byte[] newSig = Hex(e.GetProperty("new_sig_hex").GetString()!);
                outp.Add(new Identity.RotationEvidence(rec, oldAlg, oldPub, newAlg, newPub, oldSig, newSig));
            }
            return outp;
        }

        // ResolveThread: empty chain, single link, a 3-link contiguous chain, and TWO distinct
        // broken-chain shapes -- "broken_link_old_mismatch" pins the CONTIGUITY guard, and
        // "broken_link_forged_old_signature" pins the per-link CO-SIGNATURE guard (VerifyRotation),
        // isolating one guard from the other. MUTATION ANCHORS.

        [Fact]
        public void ResolveThreadMatchesOracle()
        {
            JsonElement root = LoadCorpus(out JsonDocument doc);
            using (doc)
            {
                JsonElement cases = root.GetProperty("thread").GetProperty("resolve");
                int n = 0;
                foreach (JsonElement tc in cases.EnumerateArray())
                {
                    n++;
                    string name = tc.GetProperty("name").GetString()!;
                    List<Identity.RotationEvidence> evs = BuildEvidence(tc.GetProperty("evidence"));
                    string expectError = tc.TryGetProperty("expect_error", out JsonElement ee) ? ee.GetString() ?? "" : "";

                    if (expectError.Length > 0)
                    {
                        NaalpException ex = Assert.Throws<NaalpException>(() => Identity.ResolveThread(evs));
                        Assert.True(expectError == ex.Kind, name + ": got " + ex.Kind);
                        continue;
                    }

                    Identity.Thread th = Identity.ResolveThread(evs);
                    JsonElement want = tc.GetProperty("expect_thread");
                    Assert.True(th.Root == want.GetProperty("root").GetString(), name);
                    Assert.True(th.Current == want.GetProperty("current").GetString(), name);
                    List<string> wantChain = StringList(want.GetProperty("chain"));
                    Assert.True(wantChain.Count == th.Chain.Count, name + ": chain length");
                    for (int i = 0; i < wantChain.Count; i++)
                    {
                        Assert.True(wantChain[i] == th.Chain[i], name + ": chain[" + i + "]");
                    }
                }
                Assert.True(n > 0, "no thread.resolve cases loaded");
            }
        }

        // Thread.Attributable: root/intermediate/current keys are attributable; an unrelated key is
        // not. "unrelated_key_not_attributable" is the MUTATION ANCHOR (an always-true stub flips it).

        [Fact]
        public void ThreadAttributableMatchesOracle()
        {
            JsonElement root = LoadCorpus(out JsonDocument doc);
            using (doc)
            {
                JsonElement cases = root.GetProperty("thread").GetProperty("attributable");
                int n = 0;
                foreach (JsonElement tc in cases.EnumerateArray())
                {
                    n++;
                    JsonElement thJson = tc.GetProperty("thread");
                    var th = new Identity.Thread(
                        thJson.GetProperty("root").GetString()!,
                        thJson.GetProperty("current").GetString()!,
                        StringList(thJson.GetProperty("chain")));
                    string query = tc.GetProperty("query").GetString()!;
                    bool expect = tc.GetProperty("expect").GetBoolean();
                    Assert.True(th.Attributable(query) == expect, tc.GetProperty("name").GetString());
                }
                Assert.True(n > 0, "no thread.attributable cases loaded");
            }
        }
    }
}

File: impl/java/src/test/java/sh/bubblefish/naalp/IdentityRecordsKatTest.java
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Identity RECORD + THREAD surfaces known-answer test for the Java SDK (design.md
 * §5.2/§5.3/§5.4/§5.5, R-1.4/R-5.2/R-5.3/R-5.4), graded against the shared independent corpus
 * vectors/identity_records/cases.json (tools/identity_records_oracle.py; NOT produced by this
 * code): {@link Identity.RevocationRecord}, {@link Identity#revokedAt}, {@link
 * Identity#verifyRevocation}, {@link Identity.ForeignLinkRecord}, {@link
 * Identity#verifyForeignLink}, {@link Identity.RotationEvidence}, {@link Identity.Thread}, {@link
 * Identity.Thread#attributable}, {@link Identity#resolveThread}. Go (impl/go/identity) and Rust
 * (impl/rust/src/identity.rs) MUST reproduce every *_hex byte and every expect_valid /
 * expect_linked / expect_error / expect_revoked / expect_thread / expect verdict; this test grades
 * the Java port against the SAME oracle bytes and verdicts.
 *
 * <p>SECURITY-CRITICAL: {@link Identity#verifyRevocation} is fail-closed -- a candidate signer
 * that recomputes to neither the revoked key nor a member of the deployer's configured
 * recovery-id set is rejected SignerMismatch BEFORE the signature is even checked. The
 * "recovery_key_not_configured_reject" case is the mutation anchor: a valid K_R signature over an
 * EMPTY authorized-recovery-ids set MUST reject; dropping the membership guard (authorized :=
 * true unconditionally) flips it to accept.
 *
 * <p>KAT convention (a standalone {@code main} that exits non-zero on any failure); named
 * "…KatTest" so the filename carries the "test" token the ten-language-parity gate indexes.
 * Compile alongside the SDK sources and RotationKatTest's sibling tests, then run with the
 * bcprov-jdk18on jar on the classpath (see RotationKatTest's javadoc for the exact commands).
 */
public final class IdentityRecordsKatTest {
    private static int fails = 0;

    private static void check(String name, String got, String want) {
        if (got.equals(want)) {
            System.out.println("  ok   " + name);
        } else {
            fails++;
            System.out.println("  FAIL " + name + "\n       got  " + got + "\n       want " + want);
        }
    }

    // ---- nesting-aware JSON access (Java has no JSON library; mirrors PolicyKatTest/FederationKatTest) ----

    private static Path findVector() {
        Path d = Path.of(System.getProperty("user.dir")).toAbsolutePath();
        for (int i = 0; i < 6 && d != null; i++) {
            Path p = d.resolve("vectors").resolve("identity_records").resolve("cases.json");
            if (Files.isRegularFile(p)) {
                return p;
            }
            d = d.getParent();
        }
        throw new AssertionError("vectors/identity_records/cases.json not found from " + System.getProperty("user.dir"));
    }

    private static int matchClose(String s, int open) {
        char oc = s.charAt(open);
        char cc = oc == '{' ? '}' : ']';
        int depth = 0;
        boolean inStr = false;
        for (int i = open; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (inStr) {
                if (ch == '\\') {
                    i++;
                } else if (ch == '"') {
                    inStr = false;
                }
                continue;
            }
            if (ch == '"') {
                inStr = true;
            } else if (ch == oc) {
                depth++;
            } else if (ch == cc && --depth == 0) {
                return i + 1;
            }
        }
        throw new AssertionError("unbalanced from " + open);
    }

    private static int afterKey(String s, String key) {
        Matcher m = Pattern.compile("\"" + Pattern.quote(key) + "\"\\s*:").matcher(s);
        if (!m.find()) {
            throw new AssertionError("key not found: " + key);
        }
        return m.end();
    }

    private static String objBlock(String s, String key) {
        int open = s.indexOf('{', afterKey(s, key));
        int close = matchClose(s, open);
        return s.substring(open + 1, close - 1);
    }

    private static String arrayBlock(String s, String key) {
        int open = s.indexOf('[', afterKey(s, key));
        int close = matchClose(s, open);
        return s.substring(open + 1, close - 1);
    }

    private static List<String> splitObjects(String arrayInner) {
        List<String> out = new ArrayList<>();
        int i = 0;
        while (true) {
            int open = arrayInner.indexOf('{', i);
            if (open < 0) {
                return out;
            }
            int close = matchClose(arrayInner, open);
            out.add(arrayInner.substring(open + 1, close - 1));
            i = close;
        }
    }

    private static String field(String scope, String key) {
        Matcher m = Pattern.compile("\"" + Pattern.quote(key) + "\"\\s*:\\s*\"([^\"]*)\"").matcher(scope);
        if (!m.find()) {
            throw new AssertionError("string key not found: " + key);
        }
        return m.group(1);
    }

    private static long intField(String scope, String key) {
        Matcher m = Pattern.compile("\"" + Pattern.quote(key) + "\"\\s*:\\s*(-?\\d+)").matcher(scope);
        if (!m.find()) {
            throw new AssertionError("int key not found: " + key);
        }
        return Long.parseLong(m.group(1));
    }

    private static boolean boolField(String scope, String key) {
        Matcher m = Pattern.compile("\"" + Pattern.quote(key) + "\"\\s*:\\s*(true|false)").matcher(scope);
        if (!m.find()) {
            throw new AssertionError("bool key not found: " + key);
        }
        return m.group(1).equals("true");
    }

    /** The string elements of the array named {@code key} (also fine for an empty {@code []}). */
    private static List<String> stringArray(String scope, String key) {
        Matcher a = Pattern.compile("\"" + Pattern.quote(key) + "\"\\s*:\\s*\\[(.*?)\\]", Pattern.DOTALL).matcher(scope);
        if (!a.find()) {
            throw new AssertionError("array key not found: " + key);
        }
        List<String> out = new ArrayList<>();
        Matcher m = Pattern.compile("\"([^\"]*)\"").matcher(a.group(1));
        while (m.find()) {
            out.add(m.group(1));
        }
        return out;
    }

    private static byte[] hexField(String scope, String key) {
        return Hex.decode(field(scope, key));
    }

    /** {@code errKind}-style runner: "no-error" on success, else the NaalpException.kind, else the
     * exception class name. */
    private static String errKind(Runnable r) {
        try {
            r.run();
            return "no-error";
        } catch (NaalpException e) {
            return e.kind;
        } catch (Throwable t) {
            return t.getClass().getSimpleName();
        }
    }

    // ---- revocation ----

    private static void runRevocation(String json) {
        String rev = objBlock(json, "revocation");

        // RevocationRecord.bytes() reproduces the oracle's deterministic-CBOR body.
        for (String tc : splitObjects(arrayBlock(rev, "record_bytes"))) {
            String name = field(tc, "name");
            Identity.RevocationRecord r = new Identity.RevocationRecord(field(tc, "key"), intField(tc, "not_after"));
            check("revocation.record_bytes " + name, Hex.encode(r.bytes()), field(tc, "bytes_hex"));
        }

        // RevokedAt: `posTime > notAfter`, selecting only the scenario's revocation for query_key
        // (mirrors the Go/Rust oracle test's thin selection loop over a per-key revocation set).
        for (String sc : splitObjects(arrayBlock(rev, "revoked_at"))) {
            String name = field(sc, "name");
            String queryKey = field(sc, "query_key");
            long queryPosition = intField(sc, "query_position");
            boolean expectRevoked = boolField(sc, "expect_revoked");
            boolean revoked = false;
            long notAfter = 0L;
            for (String rv : splitObjects(arrayBlock(sc, "revocations"))) {
                String key = field(rv, "key");
                if (!key.equals(queryKey)) {
                    continue;
                }
                long na = intField(rv, "not_after");
                Identity.RevocationRecord rec = new Identity.RevocationRecord(key, na);
                if (Identity.revokedAt(rec, queryPosition)) {
                    revoked = true;
                    notAfter = na;
                }
            }
            check("revocation.revoked_at " + name, Boolean.toString(revoked), Boolean.toString(expectRevoked));
            if (expectRevoked) {
                check("revocation.revoked_at " + name + " not_after", Long.toString(notAfter),
                        Long.toString(intField(sc, "expect_not_after")));
            }
        }

        // VerifyRevocation (§5.3, §5.5) -- fail-closed membership-before-signature. The mutation
        // anchor is "recovery_key_not_configured_reject": a valid K_R signature over an EMPTY
        // authorized_recovery_ids set MUST reject SignerMismatch.
        List<String> verifyCases = splitObjects(arrayBlock(rev, "verify"));
        check("revocation.verify has >= 7 cases", Boolean.toString(verifyCases.size() >= 7), "true");
        for (String tc : verifyCases) {
            String name = field(tc, "name");
            String recordScope = objBlock(tc, "record");
            Identity.RevocationRecord r = new Identity.RevocationRecord(field(recordScope, "key"), intField(recordScope, "not_after"));
            int alg = (int) intField(tc, "candidate_alg");
            byte[] pub = hexField(tc, "candidate_pubkey_hex");
            byte[] sig = hexField(tc, "sig_hex");
            List<String> recoveryIds = stringArray(tc, "authorized_recovery_ids");
            boolean expectValid = boolField(tc, "expect_valid");
            String expectKind = field(tc, "expect_error_kind");
            String got = errKind(() -> Identity.verifyRevocation(r, alg, pub, sig, recoveryIds));
            if (expectValid) {
                check("revocation.verify " + name, got, "no-error");
            } else if (!expectKind.isEmpty()) {
                check("revocation.verify " + name, got, expectKind);
            } else {
                check("revocation.verify " + name + " rejected", Boolean.toString(!got.equals("no-error")), "true");
            }
        }
    }

    // ---- foreign link ----

    private static void runForeignLink(String json) {
        String fl = objBlock(json, "foreign_link");

        // ForeignLinkRecord.bytes() reproduces the oracle's deterministic-CBOR body, and NFC/NFD
        // forms of the SAME logical foreign_id encode to DIFFERENT bytes (the mutation anchor: a
        // normalizer that collapses NFC/NFD would flip this not-equal assertion).
        String nfcBytes = null, nfdBytes = null;
        for (String tc : splitObjects(arrayBlock(fl, "record_bytes"))) {
            String name = field(tc, "name");
            Identity.ForeignLinkRecord r = new Identity.ForeignLinkRecord(
                    field(tc, "controls"), field(tc, "foreign_id"), intField(tc, "not_after"));
            String got = Hex.encode(r.bytes());
            check("foreign_link.record_bytes " + name, got, field(tc, "bytes_hex"));
            if (name.equals("nfc_form")) {
                nfcBytes = got;
            } else if (name.equals("nfd_form_different_bytes")) {
                nfdBytes = got;
            }
        }
        if (nfcBytes != null && nfdBytes != null) {
            check("foreign_link NFC/NFD encode to different bytes", Boolean.toString(!nfcBytes.equals(nfdBytes)), "true");
        }

        // VerifyForeignLink (§5.4, §5.5): valid+unexpired, the not_after boundary, expiry (ignored,
        // no error), wrong-key (ignored, no error -- same bucket as expiry), non-NFC foreign_id
        // (NonNFC, checked before expiry/signature).
        for (String tc : splitObjects(arrayBlock(fl, "verify"))) {
            String name = field(tc, "name");
            String recordScope = objBlock(tc, "record");
            Identity.ForeignLinkRecord r = new Identity.ForeignLinkRecord(
                    field(recordScope, "controls"), field(recordScope, "foreign_id"), intField(recordScope, "not_after"));
            int alg = (int) intField(tc, "candidate_alg");
            byte[] pub = hexField(tc, "candidate_pubkey_hex");
            byte[] sig = hexField(tc, "sig_hex");
            long now = intField(tc, "now");
            String expectKind = field(tc, "expect_error_kind");
            if (!expectKind.isEmpty()) {
                check("foreign_link.verify " + name, errKind(() -> Identity.verifyForeignLink(r, alg, pub, sig, now)), expectKind);
                continue;
            }
            boolean expectLinked = boolField(tc, "expect_linked");
            boolean linked;
            try {
                linked = Identity.verifyForeignLink(r, alg, pub, sig, now);
            } catch (NaalpException e) {
                fails++;
                System.out.println("  FAIL " + name + " unexpected exception " + e.kind);
                continue;
            }
            check("foreign_link.verify " + name, Boolean.toString(linked), Boolean.toString(expectLinked));
            if (expectLinked) {
                check("foreign_link.verify " + name + " controls", r.controls, field(tc, "expect_controls"));
                check("foreign_link.verify " + name + " foreign_id", r.foreignId, field(tc, "expect_foreign_id"));
            }
        }
    }

    // ---- rotation evidence / thread / resolveThread ----

    private static List<Identity.RotationEvidence> buildEvidence(List<String> evScopes) {
        List<Identity.RotationEvidence> out = new ArrayList<>();
        for (String e : evScopes) {
            Identity.RotationRecord rec = new Identity.RotationRecord(field(e, "old"), field(e, "new"), intField(e, "not_before"));
            check("thread evidence record bytes (RotationEvidence input)", Hex.encode(rec.bytes()), field(e, "record_bytes_hex"));
            out.add(new Identity.RotationEvidence(
                    rec,
                    (int) intField(e, "old_alg"), hexField(e, "old_pubkey_hex"),
                    (int) intField(e, "new_alg"), hexField(e, "new_pubkey_hex"),
                    hexField(e, "old_sig_hex"), hexField(e, "new_sig_hex")));
        }
        return out;
    }

    private static void runThread(String json) {
        String th = objBlock(json, "thread");

        // ResolveThread: empty chain, single link, a 3-link contiguous chain, and two distinct
        // broken-chain shapes -- "broken_link_old_mismatch" pins the CONTIGUITY guard and
        // "broken_link_forged_old_signature" pins the per-link CO-SIGNATURE guard, isolating one
        // guard from the other.
        for (String tc : splitObjects(arrayBlock(th, "resolve"))) {
            String name = field(tc, "name");
            List<String> evScopes = splitObjects(arrayBlock(tc, "evidence"));
            List<Identity.RotationEvidence> evs = buildEvidence(evScopes);
            String expectError = field(tc, "expect_error");
            if (!expectError.isEmpty()) {
                check("thread.resolve " + name, errKind(() -> Identity.resolveThread(evs)), expectError);
                continue;
            }
            Identity.Thread t;
            try {
                t = Identity.resolveThread(evs);
            } catch (NaalpException e) {
                fails++;
                System.out.println("  FAIL " + name + " unexpected exception " + e.kind);
                continue;
            }
            String wantScope = objBlock(tc, "expect_thread");
            check("thread.resolve " + name + " root", t.root, field(wantScope, "root"));
            check("thread.resolve " + name + " current", t.current, field(wantScope, "current"));
            check("thread.resolve " + name + " chain", String.join(",", t.chain), String.join(",", stringArray(wantScope, "chain")));
        }

        // Thread.attributable: root/intermediate/current keys are attributable; an unrelated key is
        // not ("unrelated_key_not_attributable" is the mutation anchor -- an always-true stub flips it).
        for (String tc : splitObjects(arrayBlock(th, "attributable"))) {
            String name = field(tc, "name");
            String thScope = objBlock(tc, "thread");
            Identity.Thread t = new Identity.Thread(field(thScope, "root"), field(thScope, "current"), stringArray(thScope, "chain"));
            String query = field(tc, "query");
            boolean expect = boolField(tc, "expect");
            check("thread.attributable " + name, Boolean.toString(t.attributable(query)), Boolean.toString(expect));
        }
    }

    public static void main(String[] args) throws Exception {
        String json = Files.readString(findVector(), StandardCharsets.UTF_8);
        runRevocation(json);
        runForeignLink(json);
        runThread(json);
        System.out.println(fails == 0 ? "PASS" : "FAIL (" + fails + ")");
        if (fails != 0) {
            System.exit(1);
        }
    }
}
